### ISSUE TITLE
fontTools “actions”

### ISSUE THREAD

**Issue by twardoch** [fonttools/fonttools#1869]
Apps like svgo (for SVG optimization), and many Markdown processors, have “plugins”. And of course editors like SublimeText or VSCode.

The general idea is that your processor defines a very simple API that it expects your plugin to implement, and you're then free to provide an implementation. 

fontbakery provides an extensive model for such plugins, and it's great. But I'm thinking — perhaps for the SFNT functionality of fontTools, we could define a very minimalistic "TTFontAction" class that would take a TTFont object, plus some parameters, have some very simple methods and somehow return the object. 

And it would have some human-readable name (ideally two-parted, so it could be presented in a two-tiered menu) , and maybe some method that would implement a check whether the action can be performed on a given font. 

There probably should be another class TTFontGroupAction that would take a list of TTFont objects.

Some optional knowledge of a working folder would also be nice. Or perhaps there should be a way to pass either the path or the object? See, I'm not sure. 

In a way, I'm thinking that it might be a little bit like the pens. 

Then others could adapt their own inconsistent tools that don't have a uniform API but often "do things" to a TTFont object. It might be: "freeze features", or "optimize SVG table" or "dump FEA features", or "run fontbakery" or "generate HTML proof". 

Those actions could be chained of course. A tool such as my simple [export script for FontLab 7](https://github.com/twardoch/twardoch-fontlab-scripts) could let users add these actions via a little list control. The same might actually be done in an app like FontGoggles, or maybe in a little standalone GUI app that also does SFNT table viewing/editing.

And of course there might be a commandline tool that could run those actions, with configs potentially written as a YAML/JSON.

Basically — a SIMPLE API for "snippets".

**Comment by twardoch** [comment]
I'm thinking generally three cardinalities of action classes: 

- TTAction : takes a single TTFont; could be for simple checks or fixes

- TTActionTwo : takes two TTFont objects ("new" and "old"); this would be for cases where you want to diff or to copy some table from a specially-processed reference font to your current one, for merging, for substituting "to match" etc.

- TTActionMulti : takes n TTFont objects (if needed, the 1st one is more significant); this could be done for consistency checks within a family, or comparing a VF with a bunch of statics, or merging several into one, or distributing some table to the other ones

A runner would run a series of actions of the same cardinality on a series of supplied font objects, and would ultimately do whatever it wants with it (save them back in the original location, or somewhere else)

As I said, there could be a library runner, a commandline runner, and this could go so the way to GUI runners of different kinds.
